src/vm/checker/typecheck/mod.rs

```diff
@@ -0,0 +1,404 @@
+        None
+    }
+}
+
+fn no_type() -> TypeSignature {
+    TypeSignature::new_atom(AtomTypeIdentifier::NoType)
+}
+
+fn check_atomic_type(atom: AtomTypeIdentifier, to_check: &TypeSignature) -> CheckResult<()> {
+    let expected = TypeSignature::new_atom(atom);
+    if !expected.admits_type(to_check) {
+        Err(CheckError::new(CheckErrors::TypeError(expected, to_check.clone())))
+    } else {
+        Ok(())
+    }
+}
+
+impl <'a> TypeChecker <'a> {
+    pub fn new(db: &'a AnalysisDatabase) -> TypeChecker {
+        TypeChecker {
+            db: db,
+            contract_context: ContractContext::new(),
+            type_map: TypeMap::new()
+        }
+    }
+
+    fn type_check_all(&mut self, args: &[SymbolicExpression], context: &TypingContext) -> CheckResult<Vec<TypeSignature>> {
+        let mut result = Vec::new();
+        for arg in args.iter() {
+            // don't use map here, since type_check has side-effects.
+            result.push(self.type_check(arg, context)?)
+        }
+        Ok(result)
+    }
+
+    fn type_check_function_type(&mut self, func_type: &FunctionType,
+                                args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+        let typed_args = self.type_check_all(args, context)?;
+        func_type.check_args(&typed_args)?;
+        Ok(func_type.return_type().clone())
+    }
+
+    fn type_check_list_pairs<'b> (&mut self, bindings: &[SymbolicExpression],
+                                  context: &'b TypingContext) -> CheckResult<TypingContext<'b>> {
+        let mut out_context = context.extend()?;
+        for binding in bindings.iter() {
+            let binding_exps = binding.match_list()
+                .ok_or(CheckError::new(CheckErrors::BadSyntaxBinding))?;
+            
+            if binding_exps.len() != 2 {
+                return Err(CheckError::new(CheckErrors::BadSyntaxBinding))
+            }
+
+            let var_name = binding_exps[0].match_atom()
+                .ok_or(CheckError::new(CheckErrors::BadSyntaxBinding))?;
+
+            self.type_map.set_type(&binding_exps[0], no_type())?;
+            let typed_result = self.type_check(&binding_exps[1], context)?;
+            out_context.variable_types.insert(var_name.clone(),
+                                              typed_result);
+        }
+
+        Ok(out_context)
+    }
+
+    fn get_function_type(&self, function_name: &str) -> Option<FunctionType> {
+        if let Some(function_type) = self.contract_context.get_function_type(function_name) {
+            Some(function_type.clone())
+        } else {
+            None
+        }
+    }
+
+    fn type_check_define_function(&mut self, function_expression: &[SymbolicExpression],
+                                  context: &TypingContext) -> CheckResult<(String, FunctionType)> {
+        if function_expression.len() != 3 {
+            return Err(CheckError::new(CheckErrors::IncorrectArgumentCount(2, function_expression.len() - 1)))
```

Comment:
I found this error confusing when the VM interpreter spit it out.  Would it be possible to expand all instances of `IncorrectArgumentCount` to also provide (1) the definition of the function in question so the programmer can see what arguments it expects, and (2) the invocation that didn't pass in the requisite arguments?